File: backend/app/models/schemas/device.py
# -*- coding: utf-8 -*-
"""
设备 Schema 模块

功能说明：
    定义设备相关 API 的请求和响应数据结构。
"""

from typing import Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field


class DeviceCreate(BaseModel):
    """
    创建设备请求
    
    添加新设备时需要提供的信息。
    """
    
    device_id: str = Field(
        ..., 
        min_length=1, 
        max_length=64, 
        description="设备唯一标识，用于MQTT通信"
    )
    name: str = Field(
        ..., 
        min_length=1, 
        max_length=100, 
        description="设备名称"
    )
    type: str = Field(
        ..., 
        min_length=1, 
        max_length=50, 
        description="设备类型，如 purifier、fish_feeder"
    )
    location: Optional[str] = Field(
        default=None, 
        max_length=100, 
        description="设备位置"
    )
    description: Optional[str] = Field(
        default=None, 
        description="设备描述"
    )
    config: Optional[dict] = Field(
        default=None, 
        description="设备配置"
    )


class DeviceUpdate(BaseModel):
    """
    更新设备请求
    
    修改设备信息时可以更新的字段，所有字段都是可选的。
    """
    
    name: Optional[str] = Field(
        default=None, 
        min_length=1, 
        max_length=100, 
        description="设备名称"
    )
    location: Optional[str] = Field(
        default=None, 
        max_length=100, 
        description="设备位置"
    )
    description: Optional[str] = Field(
        default=None, 
        description="设备描述"
    )
    config: Optional[dict] = Field(
        default=None, 
        description="设备配置"
    )


class DeviceResponse(BaseModel):
    """
    设备响应
    
    返回给前端的设备信息。
    """
    
    id: int = Field(description="数据库主键ID")
    device_id: str = Field(description="设备唯一标识")
    name: str = Field(description="设备名称")
    type: str = Field(description="设备类型")
    location: Optional[str] = Field(description="设备位置")
    is_online: bool = Field(description="是否在线")
    status: Optional[dict] = Field(description="设备状态")
    config: Optional[dict] = Field(description="设备配置")
    description: Optional[str] = Field(description="设备描述")
    created_at: datetime = Field(description="创建时间")
    updated_at: datetime = Field(description="更新时间")
    
    class Config:
        # 允许从 ORM 对象直接转换
        from_attributes = True


class DeviceListResponse(BaseModel):
    """
    设备列表响应
    
    分页查询设备列表的响应。
    """
    
    items: list[DeviceResponse] = Field(description="设备列表")
    total: int = Field(description="总数量")
    page: int = Field(description="当前页码")
    size: int = Field(description="每页数量")


class DeviceControlRequest(BaseModel):
    """
    设备控制请求
    
    发送控制指令到设备。
    """
    
    command: str = Field(
        ..., 
        description="控制命令，如 power_on、power_off、set_mode"
    )
    params: Optional[dict] = Field(
        default=None, 
        description="命令参数"
    )

File: admin-web/src/api/device.ts
/**
 * 设备相关 API
 * 
 * 功能说明：
 *   封装设备管理相关的接口调用。
 */

import request from './index'

// 设备类型定义
export interface Device {
  id: number
  device_id: string
  name: string
  type: string
  location?: string
  is_online: boolean
  status?: Record<string, any>
  config?: Record<string, any>
  created_at: string
  updated_at: string
}

// 创建设备参数
export interface CreateDeviceParams {
  device_id: string
  name: string
  type: string
  location?: string
  config?: Record<string, any>
}

// 更新设备参数
export interface UpdateDeviceParams {
  name?: string
  location?: string
  config?: Record<string, any>
}

// 设备列表响应
export interface DeviceListResponse {
  items: Device[]
  total: number
  page: number
  size: number
}

/**
 * 获取设备列表
 */
export function getDevices(params?: {
  page?: number
  size?: number
  type?: string
  is_online?: boolean
}): Promise<DeviceListResponse> {
  return request.get('/devices', { params })
}

/**
 * 获取设备详情
 */
export function getDevice(deviceId: string): Promise<Device> {
  return request.get(`/devices/${deviceId}`)
}

/**
 * 创建设备
 */
export function createDevice(data: CreateDeviceParams): Promise<Device> {
  return request.post('/devices', data)
}

/**
 * 更新设备
 */
export function updateDevice(deviceId: string, data: UpdateDeviceParams): Promise<Device> {
  return request.put(`/devices/${deviceId}`, data)
}

/**
 * 删除设备
 */
export function deleteDevice(deviceId: string): Promise<void> {
  return request.delete(`/devices/${deviceId}`)
}

/**
 * 控制设备
 */
export function controlDevice(deviceId: string, command: string, params?: Record<string, any>): Promise<void> {
  return request.post(`/devices/${deviceId}/control`, { command, params })
}
